### PR TITLE
Linearizer -> Lowerer

### PR DESCRIPTION
**Pull request by geohot**
- [x] Unroll
- [x] Upcast
- [x] Group
- [x] Tensor Cores
  - [x] HIP Tensor Cores
  - [x] CUDA Tensor Cores 
- [x] float4
  - [x] Image

### CODE REVIEW

**Comment by geohot** [comment]
```
FAILED test/test_linearizer.py::TestKernelOpts::test_atomic_store_multireduce - AssertionError: 
FAILED test/test_linearizer.py::TestKernelOpts::test_double_reduce_multireduce - AssertionError:                                                                                                
FAILED test/test_linearizer.py::TestKernelOpts::test_padto_sum_multireduce - AssertionError:    
FAILED test/test_linearizer.py::TestKernelOpts::test_atomic_store_nested_range_multireduce - AssertionError:                                                                                    
FAILED test/test_linearizer.py::TestKernelOpts::test_matmul_multireduce - AssertionError:       
FAILED test/test_linearizer.py::TestFloat4::test_float4_multidim_sometimes_unaligned - AssertionError: can't vectorize dtypes._float4 with size 4                                               
FAILED test/test_linearizer.py::TestFloat4::test_float4_sometimes_unaligned - assert (2, 0) == (0, 0)                                                                                           
FAILED test/test_linearizer.py::TestLinearizer::test_phi_simplification - AssertionError: has either specials or ranges, not both
FAILED test/test_linearizer.py::TestLinearizer::test_multireduce_upcasting - AssertionError: 26 != 25                                                                                           
FAILED test/test_linearizer.py::TestLinearizer::test_softmax_multireduce - AssertionError:      
FAILED test/test_linearizer.py::TestLinearizer::test_skip_unmatching_upcasts - assert (<UOps.LOAD: 19> is <UOps.CAST: 14>)                                                                      
FAILED test/test_linearizer.py::TestLinearizer::test_grouped_store_phis - AssertionError: assert (dtypes.float == dtypes._float4                                                                
FAILED test/test_linearizer.py::TestLinearizer::test_grouped_store_values - assert (dtypes._float4 == dtypes._float4                                                                            
FAILED test/test_linearizer.py::TestLinearizer::test_multireduce_store_locals - AssertionError: 2 != 3 : should have generated 3 BufferOps.STORE to the local buf but got 2                     
FAILED test/test_linearizer.py::TestLinearizer::test_tensor_cores_multi_reduce - AssertionError: tensor core not triggered                                                                      
FAILED test/test_linearizer.py::TestLinearizer::test_skip_unmatching_upcasts_with_gep - assert (<UOps.LOAD: 19> is <UOps.CAST: 14>)                                                             
FAILED test/test_linearizer.py::TestLinearizer::test_early_end_local - AssertionError: 2 != 3   
FAILED test/test_linearizer.py::TestLinearizer::test_multireduce_unroll - AssertionError: 
FAILED test/test_linearizer.py::TestLinearizer::test_tensor_cores_padded - AssertionError: tensor core not triggered                                                                            
FAILED test/test_linearizer.py::TestLinearizer::test_grouped_store_locals_and_globals - AssertionError: assert (dtypes.float == dtypes._float2                                                  
FAILED test/test_linearizer.py::TestLinearizer::test_mean_std_multireduce_multiout - tinygrad.device.CompileError: Error Domain=MTLLibraryErrorDomain Code=3 "program_source:14:16: error: use o
f undeclared identifier 'alu0'                                                                  
FAILED test/test_linearizer.py::TestLinearizer::test_reduceops_order - AssertionError: 
FAILED test/test_linearizer.py::TestLinearizer::test_arg_acc_dtype - AssertionError: assert dtypes.half == dtypes.float                                                                         
FAILED test/test_linearizer.py::TestLinearizer::test_upcast_multireduce_nested_local_upcast - AssertionError:                                                                                   
FAILED test/test_linearizer.py::TestLinearizer::test_tensor_cores - AssertionError: tensor core not triggered                                                                                   
FAILED test/test_linearizer.py::TestLinearizer::test_softmax_multireduce_multiout - AssertionError:                                                                                             
FAILED test/test_linearizer.py::TestLinearizer::test_multireduce_loop_scope - AssertionError: UOps.RANGE          : dtypes.int                [<UOps.CONST: 9>, <UOps.CONST: 9>] (2, True) not f
ound in {UOps.CONST          : dtypes.int                []              ...                    
FAILED test/test_linearizer.py::TestLinearizer::test_mean_std_multireduce_mid_dim - AssertionError:                                                                                             
FAILED test/test_linearizer.py::TestLinearizer::test_sum_acc_dtype - AssertionError: assert dtypes.half == dtypes.float                                                                         
FAILED test/test_linearizer.py::TestLinearizer::test_end_local - AssertionError: <UOps.STORE: 20> != <UOps.ENDIF: 26>     
```

File: tinygrad/codegen/lowerer.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    for i,g in enumerate(self.full_shape[self.first_reduce+self.group_for_reduces:], start=self.first_reduce+self.group_for_reduces):
+      unrolled, is_reduce = i >= (self.shape_len-self.upcasted), self.full_shape[i] != self.output_shape[i]
+      if unrolled:
+        assert isinstance(g, int), "needs to be int to unroll"
+        uop = UOp(UOps.EXPAND, dtypes.int32, tuple(UOp.const(dtypes.int32, j) for j in range(0, g)), i)
+      else:
+        uop = UOp(UOps.RANGE, dtypes.int32, (UOp.const(dtypes.int32, 0), variable_to_uop(g)), (i, is_reduce))
+      self.idxs.append(uop)
+
+    # late indexes
```
Comment (geohot) [inline tinygrad/codegen/lowerer.py:198]: This is probably generic with the reduce, it doesn't belong here

**Comment by geohot** [comment]
Two current blockers on this:

Late IF statement generation -- #5027
Multireduce fix to remove implicit broadcasting @Qazalin

A lot of my struggles yesterday were actually the implicit broadcasting not being present. With that fixed, I think the lowerer can pass all tests, then it's a question of speed.

**Review by Qazalin — COMMENTED** [review]
hmm `test_softmax_multireduce_multiout` is still wrong but verify_lazyop didn't catch it.

**Comment by geohot** [comment]
Tomorrow, fix `GRAPHUOPS=1 GPU=1 IMAGE=2 FORWARD_ONLY=1 DEBUG=5 python3 test/test_ops.py TestOps.test_simple_conv2d`

**Review by Qazalin — COMMENTED** [review]
merged late IF.
Upcasts (n)x the barrier (eg. `test/test_linearizer.py TestLinearizer.test_grouped_store_locals_and_globals`) - I'm working on late BARRIER next to delete it from the Lowerer.


also I think L116 is wrong for multi output:
```
if self.group_for_reduces > 0 and x.arg.idx == 0: valid, has_valid = valid * self.idxs[self.first_reduce].eq(0), True
```
`test/test_assign.py TestAssign.test_permuted_reduceop_multioutput_dual_use_possible`

**Comment by geohot** [comment]
LLVM + CLANG + GPU passing!

**Comment by geohot** [comment]
All that's left is non METAL tensor cores (and a lil bug in metal ones)

This will be a tomorrow thing, then speedups on Wednesday, then merge!

File: tinygrad/codegen/uops.py [hunk 2]
```diff
@@ -200,16 +203,241 @@ def sum_collapse(phi_input, loop, val1, val2):
...
+        if cc in replace:
+          # NOTE: handle expands that are already replaced
+          tcc = replace[cc]
+          replace[cc] = UOp(tcc.op, tcc.dtype, tuple(replace.get(x, x) for x in tcc.src), tcc.arg)
+        else:
+          replace[cc] = UOp(cc.op, cc.dtype, tuple(replace.get(x, x) for x in cc.src), cc.arg)
+
+  return [x.get(base, base) for x in replaces]
+
+def get_reduce_acc(op, dtype):
```
Comment (chenyuxyz) [inline tinygrad/codegen/uops.py:287]: this has been refactored in linearizer
```
  def get_reduce_acc(self, reduceop:LazyOp):
    if reduceop.op is ReduceOps.SUM: return dtypes.as_const(0, reduceop.dtype)
    if reduceop.op is ReduceOps.MAX: return dtypes.min(reduceop.dtype)
```

**Comment by Qazalin** [comment]
- PTX fix: 2a42c637b99d5fcad7952c0cae346d09cbb5c0af

-----


- How are some of the ASTs from fixup_ast valid?

`DEBUG=4 python3 test/test_schedule.py TestSchedule.test_argmax_multireduce_fusion` with 6c28178341eae5c1a604a7d334204ac3aafeaf0f:

```
r_4_16_2
  0 ━┳ BufferOps.STORE MemBuffer(idx=0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(4, 1, 1), strides=(1, 0, 0), offset=0, mask=None, contiguous=True),)))
  1  ┗━┳ ReduceOps.MAX (1,)
  2    ┗━┳ BufferOps.LOAD MemBuffer(idx=-1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(1, 16, 1), strides=(0, 1, 0), offset=0, mask=None, contiguous=True),)))
  3      ┗━┳ BufferOps.STORE MemBuffer(idx=-1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(1, 16, 1), strides=(0, 1, 0), offset=0, mask=None, contiguous=True),)))
  4        ┗━┳ ReduceOps.MAX (2,)
  5          ┗━━ BufferOps.LOAD MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(4, 16, 2), strides=(32, 2, 1), offset=0, mask=None, contiguous=True),)))
  

AssertionError: found implicit movement op ReduceOps.MAX (4, 16, 1) != BufferOps.STORE (1, 16, 1)
```

**Comment by geohot** [comment]
Thanks for the PTX fix!

Hmm, fair. It's valid because that LOAD/STORE are to locals, so the global can be 1. But yea....there's not really a way to check that given only the LazyOp AST.

**Comment by Qazalin** [comment]
Here is a CI run of all the cases where fixed up LazyOp is invalid:
https://github.com/tinygrad/tinygrad/actions/runs/9870064451

File: tinygrad/codegen/uops.py [hunk 3]
```diff
@@ -394,12 +644,30 @@ def _dfs(u:UOp, gate:UOp) -> UOp:
       if (replace_source:=tuple(_dfs(x, gate) for x in u.src)) != u.src: return UOp(u.op, u.dtype, replace_source, u.arg)
       return u
     for i, s in enumerate(self.sinks[:]):
-      if s.op is UOps.STORE and len(s.src) == 4 and (rw:=_dfs(s, s.src[3])) != s: self.sinks[i] = UOp(rw.op, rw.dtype, rw.src[:3], rw.arg)
+      # breaks for TC. maybe move WMMA to the lowerer?
+      if all(x.op is not UOps.TC for x in s.parents):
```
Comment (flammit) [inline tinygrad/codegen/uops.py:653]: I might be missing something, but is this still needed?

Comment (geohot) [reply tinygrad/codegen/uops.py:653]: ohhh, that's actually a bug now I think, but it's WMMA, not TC!

**Comment by geohot** [comment]
`DEBUG=5 NV=1 M_START=9 M_STOP=9 M_STEP=1 N_START=9 N_STOP=9 N_STEP=1 K_START=5 K_STOP=5 K_STEP=1 HALF=1 TC_OPT=2 python3 ./extra/gemm/fuzz_matmul.py` is failing

**Pull request merged**